[shelly] Add support for BLU Wallswitch 4, RC Button 4, Remote, Tough ZB, H&T ZB, BLU Distance

markus7017 opened the pull request:
This PR is based on work of @udo1toni, who did the initial implementation, see #18841
Due to a high number of rebase conflicts I decided to create a new PR and finalize the implementation also adopting last refactoring changes.

Mainly the transportation scheme from a device used as BLU gateway to OH shelly binding is adapted:
- the BLU gateway script oh-blu-scanner.js is extended so that it first collects all 0x3A events in an array befor adding to the JSON serializer. Script has been optimized using ChatGPT (will receive more updates with integration of new Shelly devices)
- JSON serialization class Shelly2NotifyEventMessage: member Button changed to Buttons, instead of a single byte now a byte array is transported where the index ([0..3]) within the array represents the button ID [1..4]
- Thing initialization now uses full model ID for the different buttons to support BLU Button as well as the multi button devices
- A new flag isMultiButton was added to differ between single buttons (isButton) and multi buttons like Wallswitch 4
- A new state HOLD is reported, but doesn't triggers the event count, otherwise every click would increment the counter twice, now maps both states (254 and 128 from new firmware) to state HOLD

In addition the PR adds support for
- BLU Tough ZB
- BLU Wallswitch 4
- BLU RC Remote 4
- BLT H&T ZB
- BLU Distance
- BLU Remote

Awaiting PR #19093 to adapt to the refactoring changes and finalize implementation.
Awaiting PR ##19170, which refactored the BLU support before adding new devices.

@UdoKrie Many thanks for the initial work, this was on my list for a long time, but lack of JavaScript experience... Could you please review the lastest script if you see issues. 

Closing:
- #18142 
- #18361
- #19252
- #18171
- #18427

markus7017 committed:
wip

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
wip

markus7017 committed:
wip

markus7017 committed:
first working version

markus7017 committed:
wip

markus7017 committed:
state HOLDING renamed to HOLD
Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1779,7 +1783,34 @@ See notes on discovery of Shelly BLU devices above.
+
+# Shelly BLU Wall Switch 4 (thing-type: shellybluwallswitch4)
+
+ee notes on discovery of Shelly BLU devices above.
```
fixed

markus7017 committed:
BLU Tough ZB and BLU H&T ZB added

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
Update script to return Temperatur or Temperatures depending if this is
a single value or an array.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[WIP] Support for processing of Buttons, Temperature updates revised
(now always an array), Channel+Rotation+Dimmer added, varios changes.
Processing of Dimmer values is missing and BLU Distance incomplete;
README not updated

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
BLU Remote: Channels rotation1/2/3, direction, steps implemented, bug
fixing

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
wip

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
add channels for Shelly Distance

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
added firmware channel, some fixing

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
fix distance channel

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
rebase

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
clean-up, README updated

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
updated

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur reviewed (changes requested):
@markus7017 - can you please rebase and make sure no changes from #19170 are reverted? A self-review might help.

markus7017 commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1746,6 +1756,60 @@ See notes on discovery of Shelly BLU devices above.
+|         | lowBattery    | Switch   | yes       | Low battery alert (< 20%)                                                           |
+| device  | gatewayDevice | String   | yes       | Shelly forwarded last status update (BLU gateway), could vary from packet to packet |
+
+
```
fixed

markus7017 commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1746,6 +1756,60 @@ See notes on discovery of Shelly BLU devices above.
+|         | eventCount    | Number   | yes       | Counter gets incremented every time the device issues a button event.               |
+|         | button        | Trigger  | yes       | Event trigger with payload, see SHORT_PRESSED or LONG_PRESSED                       |
+|         | lastUpdate    | DateTime | yes       | Timestamp of the last measurement                                                   |
+| status2 | lastEvent     | String   | yes       | Last event type (S/SS/SSS/L/H) for button 1                                         |
```
fixed

markus7017 committed:
MAP_BLU_INPUT_EVENT_TYPE: keys changed to String again, to allow
returning an empty value (""); fix naming of input channels for multi
button devices (e.g. status1#lastEvent instead of status1#lastEvent1),
BLU Distance has no inputs (the button is inside the housing); README
updated

markus7017 committed:
suppress error message on socket disconnect after discovery (no thing
created);

markus7017 committed:
fix debug msg on incoming BLU packet; BLU Distance added back to scanner
script;

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
some fixes

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1541,6 +1547,10 @@ Channels lastEvent and eventCount are only available if input type is set to mom
 
+## Shelly BLU Gateway, BLU Gateway 3 (thing-type: shellyblugw)
+
+There are no additional channels beside the device group.
```
fixed

jlaur commented:
> where do you see code from #19170?

Example of reverted change from that PR:

<img width="1612" height="163" alt="image" src="https://github.com/user-attachments/assets/fd453f23-2fb4-4e76-9461-390c5cde1ab0" />

markus7017 committed:
review change

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented:
> > where do you see code from #19170?
> 
> Example of reverted change from that PR:

ok, fixed

fyi: 
- ShellyBluEventDataDTO moved back to Shelly2GsonDTO, because it does not make sense to have a separate file (I thought to implement more stuff there, but went a different direction)
- SHELLY2_EVENT_HOLDING back to HOLDING, because it's not a momentary state, you need to keep holding the button
- MAP_BLU_INPUT_EVENT_TYPE: key reverted to String, otherwise I need a 2nd version of mapValue() and I can't return an empty integer like an empty strong
- Processing of BLU Sensor Data optimized (cleaner, avoids NPEs)
- THING_TYPE_SHELLYBUTTON renamed to THING_TYPE_SHELLYBUTTON1, because that's the name of the device
- more unit tests added
- Added for devices to BLU CLASSses and BLU Device Types, implemented new channels

markus7017 committed:
MAP_BLU_INPUT_EVENT_TYPE uses Int values as key; don't fill input
channels on hold for multi button devices (firmware bug); create
distance channel also when not included in 1st packet; avoid NPE on
steps channel not filled; BLU Remote re-added to script

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur reviewed (changes requested):
Thanks for adding support for all the new stuff! I have now done a full review, please find my comments below.

jlaur commented:
> * ShellyBluEventDataDTO moved back to Shelly2GsonDTO, because it does not make sense to have a separate file (I thought to implement more stuff there, but went a different direction)

It makes perfect sense to divide into smaller files and keep one class per file.

markus7017 committed:
review changes

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
review change - create seperate file

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
change distance precision to 2 digits (device measures with 1cm
accurancy)

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiRpc.java:
```diff
@@ -491,13 +497,24 @@ protected void installScript(String script, boolean install) throws ShellyApiExc
-                ourId = rsp.id;
-                logger.debug("{}: Script has been created, id={}", thingName, ourId);
-                upload = true;
+                if (ourId == -1) {
+                    // find free script id
+                    ourId = 0;
+                    for (ourId = 1; ourId <= MAX_SCRIPT_ID && testScriptId(scriptList, ourId); ourId++) {
+                    }
+                }
+                if (ourId <= MAX_SCRIPT_ID) {
+                    // Create new script, get id
+                    ShellyScriptResponse rsp = apiRequest(new Shelly2RpcRequest()
+                            .withMethod(SHELLYRPC_METHOD_SCRIPT_CREATE).withId(ourId).withName(script),
+                            ShellyScriptResponse.class);
+                    ourId = rsp.id;
+                    logger.debug("{}: Script has been created, id={}", thingName, ourId);
+                    upload = true;
+                } else {
+                    logger.debug("{}: Too many scripts installed on this device", thingName);
+                    upload = false;
+                }
```
This is not expected in this PR as part of #19231.

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyDevices.java:
```diff
@@ -643,7 +675,12 @@ public class ShellyDevices {
             Map.entry(THING_TYPE_SHELLYBLUHT, 1), //
-            Map.entry(THING_TYPE_SHELLYBLUDW, 1));
+            Map.entry(THING_TYPE_SHELLYBLUDW, 1), //
+            Map.entry(THING_TYPE_SHELLYBLUMOTION, 0), // button is iside the housing
+            Map.entry(THING_TYPE_SHELLYBLUDISTANCE, 0), // button is iside the housing
```
```suggestion
            Map.entry(THING_TYPE_SHELLYBLUMOTION, 0), // button is inside the housing
            Map.entry(THING_TYPE_SHELLYBLUDISTANCE, 0), // button is inside the housing
```

I guess?

markus7017 committed:
Unit for distance channel changed to mmi

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
review changes

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java:
```diff
@@ -1004,16 +1005,24 @@ protected Shelly2RpcBaseMessage buildRequest(String method, @Nullable Object par
-    protected String mapValue(Map<String, String> map, @Nullable String key) {
-        if (key == null || key.isEmpty()) {
+    protected String mapValue(Map<String, String> map, String key) {
+        String value = map.getOrDefault(key, "");
+        if (value.isEmpty()) {
```
This would be more straight-forward, since I assume that the provided `Map`s don't contain empty strings as values.
```suggestion
        String value = map.get(key);
        if (value == null) {
```

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java:
```diff
@@ -1004,16 +1005,24 @@ protected Shelly2RpcBaseMessage buildRequest(String method, @Nullable Object par
+    }
+
+    protected String mapIntValue(Map<Integer, String> map, Integer key) {
+        String value = map.getOrDefault(key, "");
+        if (value.isEmpty()) {
```
See above.

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/ShellyBluApi.java:
```diff
@@ -292,23 +292,29 @@ public void onNotifyEvent(Shelly2RpcNotifyEvent message) {
+                                        deviceStatus.inputs.set(bttnIdx, input);
                                     } else {
-                                        logger.debug("{}: ignore H, pid={}", message.src, e.blu.pid);
+                                        logger.debug("{}: Unknown Button event", thingName);
```
Already logged by `mapIntValue`, really needed again? When explicitly logging here, again there is no need for `mapIntValue` at all. It could be as simple as:
```java
                                    String btnEvent = MAP_BLU_INPUT_EVENT_TYPE.get(e.blu.buttons[bttnIdx]);
                                    if (btnEvent == null) {
                                        logger.debug("{}: Unknown Button event {}", thingName, e.blu.buttons[bttnIdx]);
                                        continue;
                                    }
                                    // ignore HOLDING events for counter and trigger
```

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:
```diff
@@ -543,7 +577,9 @@ public static boolean updateSensors(ShellyThingInterface thingHandler, ShellySet
         // Update Add-On channeös
-        if (status.extTemperature != null) {
+        if (status.extTemperature != null)
+
```
```suggestion
```

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyComponents.java:
```diff
@@ -543,7 +577,9 @@ public static boolean updateSensors(ShellyThingInterface thingHandler, ShellySet
 
         // Update Add-On channeös
-        if (status.extTemperature != null) {
+        if (status.extTemperature != null)
+
+        {
```
Something weird happened here.
```suggestion
        if (status.extTemperature != null) {
```

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:
```diff
@@ -562,14 +575,24 @@ public static Map<String, Channel> createSensorChannels(final Thing thing, final
+        // In this case create both channels, also if only one is included in the first packet
+        if (sdata.distance != null || profile.isDistance) {
+            addChannel(thing, newChannels, sdata.distance != null, CHANNEL_GROUP_SENSOR, CHANNEL_SENSOR_DISTANCE);
+
```
```suggestion
```

markus7017 committed:
rotation1,2,3 are changed to X,Y,Z

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/ShellyBluApi.java:
```diff
@@ -256,37 +249,50 @@ public void onNotifyEvent(Shelly2RpcNotifyEvent message) {
+                            sensorData.sensor.motion = e.blu.motionState == 1;
+                        }
+                        if (e.blu.rotations != null) {
+                            logger.trace("{}: Shelly BLU rotation info received: {}", thingName,
+                                    gson.toJson(e.blu.rotations));
```
Reopened, nothing was done except resolving the comment without commenting.

jlaur commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api1/Shelly1ApiJsonDTO.java:
```diff
@@ -1193,6 +1194,15 @@ public static class ShellyExtSwitchStatusInput {
+        public Integer channel;
+        public String direction;
+        public Integer steps;
+        public Double rotationX;
+        public Double rotationY;
+        public Double rotationZ;
```
These were changed from rotation1..rotation3 in the DTO in latest commit. So just checking: was it broken before, or is it broken now?

markus7017 committed:
review changes

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
review changes

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
Add "Enable BLU Gateway" and "Enable Range Extender" thing config
options for Plus RGBW - see #19252

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
added separate profile for rgbw-gen2 #19252

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
fix the fix

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented:
@jlaur Igor is currently testing. BLU devices themself look good so far (included a 12h+ run to see updates to come over a linger time). Now we do some more testing with non-BLU sensors to check for side effects.

there are 2 or 3 discussions above I'm interested in your response or can be marked as resolved

finally we'll check the README for consistency

markus7017 commented:
@jlaur Igor is currently testing. BLU devices themself look good so far (included a 12h+ run to see updates to come over a linger time). Now we do some more testing with non-BLU sensors to check for side effects.

there are 2 or 3 discussions above I'm interested in your response or can be marked as resolved

finally we'll check the README for consistency

markus7017 committed:
fix equippment types for BLU Distance and BLU H&T

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
logging of dimmer event info removed (this is part of the eventJSON,
which gets logged before)

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented:
@jlaur I walked through all conversations and found 3x the gson.toJson(), which are no longer required because the complete event JSON is already logged. 

We did some additional testing with Gen1 DW2 and Motion to check for side effects in the shared code, which also looks good.

Depending on you final answers we are ready to merge.

markus7017 commented:
I checked again alls unresolved and resolved entries and made sure that all of them ended with a comment from my side. I'm using the Web View, unaided all comments, clicked "Show resolved".
What do you mean with "publish other comments"? I assume all comments are instantly visible, or how to publish pending comments?

markus7017 committed:
syntax error fixed

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur commented:
> What do you mean with "publish other comments"? I assume all comments are instantly visible, or how to publish pending comments?

See https://github.com/orgs/community/discussions/10369

markus7017 commented:
do you now see 34 comments?

jlaur commented:
> do you see now 34 comments?

Yes, they are now published. I'll have a look during the weekend.

jlaur reviewed (commented):
I think I went through all comments now, only a few left to conclude.

markus7017 committed:
review change

Signed-off-by: Markus Michels <markus7017@gmail.com>

jlaur commented on bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/device.xml:
```diff
@@ -155,4 +155,11 @@
 		</state>
 	</channel-type>
 
+	<channel-type id="deviceFirmware" advanced="true">
+		<item-type>String</item-type>
+		<label>@text/channel-type.shelly.deviceFirmware.label</label>
+		<description>@text/channel-type.shelly.firmware.description</description>
+		<state pattern="%s" readOnly="true">
+		</state>
+	</channel-type>
```
OK, let's leave it in then and hope this won't become a trend. 🙂

jlaur reviewed (approved):
LGTM
